### PR TITLE
win-capture: Add option to allow transparency

### PR DESCRIPTION
**Pull request by WamWooWam**
### Description
Adds an "Allow transparency" checkbox to Window Capture on Windows to preserve transparency of captured windows.

### Motivation and Context
Both currently supported methods of capturing windows in OBS preserve transparency at their source, however OBS explicitly disables this by rendering as opaque. This means in situations where you'd want to capture a window without it's background (e.g. LiveSplit), you'd have to do hacky workarounds like keying a colour, destroying text antialiasing & shadows, and creating awful artefacts if not done properly.

This PR allows users to preserve the transparency of their source window into OBS, with no keying nonsense and perfect quality, and also allows the correct capture of non-square windows (e.g. Windows 11)

It's important to note BitBlt and WCG both preserve transparency in slightly different ways. BitBlt works on the window's DC level, so has some odd behaviours with transparency on some windows (some colours appearing fully transparent even though they aren't), meanwhile WGC works on the compositor level and generally provides results more accurate to how the window appears on the desktop.

### How Has This Been Tested?
Tested on LiveSplit, Chrome, Windows Explorer, Task Manager, Electron and iTunes on Windows 10 21H1

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/win-capture/window-capture.c [hunk 1]
```diff
@@ -428,6 +432,9 @@ static obs_properties_t *wc_properties(void *data)
 
 	obs_properties_add_bool(ppts, "client_area", TEXT_CLIENT_AREA);
 
+	obs_properties_add_bool(ppts, "allow_transparency",
+				TEXT_ALLOW_TRANSPARENCY);
+
```
Comment (jpark37) [inline plugins/win-capture/window-capture.c:437]: Move this above "cursor" for consistency with Game Capture.

**Review by jpark37 — CHANGES_REQUESTED** [review]
When I try capturing LiveSplit with about half opacity, the captured texture still has 100% alpha for both BitBlt and WGC. Which applications were you able to get transparency captured?

**Comment by jpark37** [comment]
I just saw the Discord chat. Does this only work on Windows 11 maybe?

**Comment by WamWooWam** [comment]
> When I try capturing LiveSplit with about half opacity, the captured texture still has 100% alpha for both BitBlt and WGC. Which applications were you able to get transparency captured?

So for LiveSplit, you need to specify a transparent background colour in your layout settings, then set capture mode to BitBlt, and check the Allow Transparency checkbox, like:

image

The opacity slider does nothing to OBS' capture, I believe this to be a Windows limitation.

> I just saw the Discord chat. Does this only work on Windows 11 maybe?

All testing was done initially on Windows 10 21H1, I only tested Windows 11 to see if the rounded window corners would be captured correctly

**Comment by jpark37** [comment]
I see, so that made BitBlt work, but WGC is still opaque. Are there cases where WGC works?

Supporting semi-transparency for window capture is a bit of a nightmare because Windows does not composite in linear space, so if a user attempts to replicate Windows blending in OBS, they will fail. It wouldn't have been a problem before v27 because we were doing the same bad thing, but now that we are more correct, we can't align with bad compositors.

**Comment by WamWooWam** [comment]
> I see, so that made BitBlt work, but WGC is still opaque. Are there cases where WGC works?
For windows where they appear transparent in the compositor (so AIMP and WPF), WGC works best. Examples [here](https://discord.com/channels/348973006581923840/374636084883095554/860225518506344498) and[here](https://discord.com/channels/348973006581923840/374636084883095554/860220789957197824).

> Supporting semi-transparency for window capture is a bit of a nightmare because Windows does not composite in linear space, so if a user attempts to replicate Windows blending in OBS, they will fail. It wouldn't have been a problem before v27 because we were doing the same bad thing, but now that we are more correct, we can't align with bad compositors.

So this is a slightly odd one. It appears OBS *does* blend in linear space for `BitBlt` captures without `Multi Adapter Compatibility` enabled. 

I'm sorta in two minds on this, it is a slightly odd result that definitely defied my own expectations, but at the same time, the current behaviour is the "correct" thing to do, though maybe a checkbox a-la `Apply alpha in linear space` on Image Source?

**Comment by jpark37** [comment]
> So this is a slightly odd one. It appears OBS does blend in non-linear space for BitBlt captures without Multi Adapter Compatibility enabled.

This is temporary oversight. We are planning to make things consistent in #4950, which will cause this case to blend in linear.

> I'm sorta in two minds on this, it is a slightly odd result that definitely defied my own expectations, but at the same time, the current behaviour is the "correct" thing to do, though maybe a checkbox a-la Apply alpha in linear space on Image Source?

The image source hack only solves half of the problem. It is correct if you blend into a black background, but if you blend into any other color, the result will be brighter than expected. There's no reason to make it an option since Windows will never give us a linear-correct semi-transparent color value, so we should make window capture do the hack path if we accept this PR.

I'm not really a fan of how this seems to work in some cases and not other cases, and the LiveSplit setup is kind of crazy. (Modify the alpha in a place where Windows itself doesn't respect it, but OBS does, and it only works for one of the two capture modes.) I'd be for this PR if it was straightforward and worked reliably; it just feels too user-unfriendly, and several problems are not within our control. Even being behind a nondefault toggle, I suspect it's going to make users mad when if they try to use it, and it doesn't work as they expect.

**Comment by WamWooWam** [comment]
> I'm not really a fan of how this seems to work in some cases and not other cases, and the LiveSplit setup is kind of crazy. (Modify the alpha in a place where Windows itself doesn't respect it, but OBS does, and it only works for one of the two capture modes.) I'd be for this PR if it was straightforward and worked reliably; it just feels too user-unfriendly, and several problems are not within our control. Even being behind a nondefault toggle, I suspect it's going to make users mad if they try to use it, and it doesn't work as they expect.

So, while I absolutely agree that some of this works kinda weirdly, unfortunately it's mostly down to limitations within Windows itself to my knowledge? There may be more things we can do to improve the experience that I'm simply unaware of (I haven't taken a look at the deep inner workings of BitBlt capture for example).

That said, I do feel it makes some scenarios *significantly* easier, and allows for other much more powerful ones. While it doesn't work everywhere, where it does work, in my opinion, the benefits vastly outweigh the drawbacks. It's not like Window Capture isn't a bit funky to begin with (not capturing windows sometimes, returning blank or black surfaces, etc.), nor is it like we don't have knobs already that users probably shouldn't touch if they don't know what they're doing (see the aforementioned `Apply alpha in linear space`). 

That said, to my knowledge there are no major issues (besides colour blending) with enabling this purely for WGC, and leaving BitBlt alone. WGC pretty reliably returns a surface identical to what the compositor works with.

Also if you wanna discuss this in the Discord, do ping me, it seems a tad easier than via GitHub issues :p

**Comment by izzy** [comment]
> Outside of curved corners in Windows 11, how many apps these days don't use square windows? I remember back in the days of WPF there'd be many apps with all sorts of shapes, but cannot say I've run into any such apps in the past 5+ years.

Don't think about it like "which corners aren't square" but more like: What features can be enabled with this.

I'm currently in need of this specifically to not have to make my app's background green or some other colour but instead just make the whole window transparent and then draw the content. Since game capture doesn't work with my app, window capture with transparency support is a must for my use-case.

And I know, keying exists, but it's a lot easier(and cheaper in terms of performance) to just give the window transparency compared to having to deal with colour/chroma keying filters in OBS and the issues that come with that(like not being able to use the colour, added performance requirements of filters, etc.).

**Comment by SirMangler** [comment]
In my humble opinion, this is a feature that should have been added long ago. I wholeheartedly agree with Wam's assessment of the scenario. There is no "one size fits all" solution, primarily due to the "Windows factor", where Microsoft rarely respects it's own standards (See WinForms components completely not working in WPF Windows with transparency enabled, despite WinForms supporting transparency). 

From the POV of both a user and developer, I would really like this to be merged if possible, just because it doesn't benefit every scenario doesn't mean it won't be greatly beneficial for many others.

**Comment by WamWooWam** [comment]
Given this was solved by #6414 i'mma close this

**Pull request closed without merging**